File: services/authService.ts
/**
 * Authentication Service
 * Xử lý communication với backend auth APIs
 */

import { API_BASE_URL } from '../config/api';

console.log('📡 AuthService loaded with API_BASE_URL:', API_BASE_URL);

export interface User {
  id: string;
  email: string;
  username: string;
  role: string;
  status: string;
  email_verified: boolean;
  created_at: string;
  last_login: string;
}

export interface LoginRequest {
  email: string;
  password: string;
}

export interface RegisterRequest {
  email: string;
  username: string;
  password: string;
  confirm_password: string;
}

export interface AuthResponse {
  user: User;
  access_token: string;
  token_type: string;
  expires_in: number;
}

export interface UserProfile {
  display_name: string;
  phone: string;
  notify_email: boolean;
  notify_push: boolean;
  language: string;
  default_currency: string;
}

export interface UserDetail {
  user: User;
  profile: UserProfile;
}

export interface ProfileUpdateRequest {
  display_name?: string;
  phone?: string;
  notify_email?: boolean;
  notify_push?: boolean;
  language?: string;
  default_currency?: string;
}

export interface PasswordChangeRequest {
  current_password: string;
  new_password: string;
  confirm_new_password: string;
}

export interface LastLogin {
  last_login: string;
  time_ago: string;
  formatted: string;
}

export class AuthService {
  /**
   * Đăng ký tài khoản mới
   */
  static async register(data: RegisterRequest): Promise<AuthResponse> {
    try {
      console.log('📤 Sending register request to:', `${API_BASE_URL}/auth/register`);

      const response = await fetch(`${API_BASE_URL}/auth/register`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(data),
      });

      console.log('📥 Response status:', response.status);

      if (!response.ok) {
        let errorMessage = 'Đăng ký thất bại';
        try {
          const error = await response.json();
          errorMessage = error.detail || error.message || errorMessage;
        } catch (e) {
          errorMessage = `HTTP ${response.status}: ${response.statusText}`;
        }
        console.error('❌ Register error:', errorMessage);
        throw new Error(errorMessage);
      }

      const result: AuthResponse = await response.json();
      console.log('✅ Register success:', result.user);

      // Lưu token vào localStorage
      localStorage.setItem('access_token', result.access_token);
      localStorage.setItem('user', JSON.stringify(result.user));

      return result;
    } catch (error) {
      console.error('❌ Register fetch error:', error);
      throw error;
    }
  }

  /**
   * Đăng nhập
   */
  static async login(data: LoginRequest): Promise<AuthResponse> {
    try {
      console.log('📤 Sending login request to:', `${API_BASE_URL}/auth/login`);

      const response = await fetch(`${API_BASE_URL}/auth/login`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(data),
      });

      console.log('📥 Response status:', response.status);

      if (!response.ok) {
        let errorMessage = 'Đăng nhập thất bại';
        try {
          const error = await response.json();
          errorMessage = error.detail || error.message || errorMessage;
        } catch (e) {
          errorMessage = `HTTP ${response.status}: ${response.statusText}`;
        }
        console.error('❌ Login error:', errorMessage);
        throw new Error(errorMessage);
      }

      const result: AuthResponse = await response.json();
      console.log('✅ Login success:', result.user);
      
      // Lưu token vào localStorage
      localStorage.setItem('access_token', result.access_token);
      localStorage.setItem('user', JSON.stringify(result.user));
      
      return result;
    } catch (error) {
      console.error('❌ Login fetch error:', error);
      throw error;
    }
  }

  /**
   * Đăng xuất
   */
  static async logout(): Promise<void> {
    try {
      const token = this.getToken();
      
      if (token) {
        await fetch(`${API_BASE_URL}/auth/logout`, {
          method: 'POST',
          headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
          },
        });
      }
    } catch (error) {
      console.error('❌ Logout error:', error);
    } finally {
      // Xóa token và user data
      localStorage.removeItem('access_token');
      localStorage.removeItem('user');
      console.log('✅ Logged out');
    }
  }

  /**
   * Lấy token từ localStorage
   */
  static getToken(): string | null {
    return localStorage.getItem('access_token');
  }

  /**
   * Lấy user từ localStorage
   */
  static getUser(): User | null {
    const user = localStorage.getItem('user');
    return user ? JSON.parse(user) : null;
  }

  /**
   * Kiểm tra xem user đã đăng nhập hay chưa
   */
  static isAuthenticated(): boolean {
    return !!this.getToken();
  }

  /**
   * Lấy thông tin chi tiết user (bao gồm profile)
   */
  static async getCurrentUser(): Promise<UserDetail> {
    try {
      const token = this.getToken();
      
      if (!token) {
        throw new Error('Không có token');
      }

      const response = await fetch(`${API_BASE_URL}/auth/me`, {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
      });

      console.log('📥 GET /auth/me response status:', response.status);

      if (!response.ok) {
        throw new Error(`Không thể lấy thông tin user: HTTP ${response.status}`);
      }

      const result = await response.json();
      console.log('✅ Current user loaded:', result);

      // Backend trả về định dạng khác có thể, xử lý linh hoạt
      if (!result.user) {
        // Nếu result là user data trực tiếp
        result.user = result;
      }
      
      if (!result.profile) {
        // Tạo profile mặc định
        result.profile = {
          display_name: result.user?.username || 'User',
          phone: result.user?.phone || '',
          notify_email: true,
          notify_push: true,
          language: 'vi',
          default_currency: 'VND'
        };
      }

      return result as UserDetail;
    } catch (error) {
      console.error('❌ Get current user error:', error);
      throw error;
    }
  }

  /**
   * Cập nhật thông tin cá nhân
   * Method: PUT
   * Path: /api/users/me/profile
   */
  static async updateProfile(data: ProfileUpdateRequest): Promise<UserProfile> {
    try {
      const token = this.getToken();
      
      if (!token) {
        throw new Error('Không có token');
      }

      console.log('📤 Sending PUT request to:', `${API_BASE_URL}/users/me/profile`);

      const response = await fetch(`${API_BASE_URL}/users/me/profile`, {
        method: 'PUT',
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(data),
      });

      console.log('📥 Response status:', response.status);

      if (!response.ok) {
        let errorMessage = 'Cập nhật profile thất bại';
        try {
          const error = await response.json();
          errorMessage = error.detail || error.message || errorMessage;
        } catch (e) {
          errorMessage = `HTTP ${response.status}: ${response.statusText}`;
        }
        console.error('❌ Update profile error:', errorMessage);
        throw new Error(errorMessage);
      }

      const result = await response.json();
      console.log('✅ Profile updated:', result);

      return result;
    } catch (error) {
      console.error('❌ Update profile fetch error:', error);
      throw error;
    }
  }

  /**
   * Đổi mật khẩu
   */
  static async changePassword(data: PasswordChangeRequest): Promise<{ message: string }> {
    try {
      const token = this.getToken();
      
      if (!token) {
        throw new Error('Không có token');
      }

      const response = await fetch(`${API_BASE_URL}/auth/change-password`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(data),
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || error.message || 'Đổi mật khẩu thất bại');
      }

      const result = await response.json();
      console.log('✅ Password changed:', result);

      return result;
    } catch (error) {
      console.error('❌ Change password error:', error);
      throw error;
    }
  }

  /**
   * Lấy thông tin đăng nhập lần cuối
   * Path: GET /api/auth/last-login
   */
  static async getLastLogin(): Promise<LastLogin | null> {
    try {
      const token = this.getToken();
      
      if (!token) {
        throw new Error('Không có token');
      }

      console.log('📤 Fetching last login info from:', `${API_BASE_URL}/auth/last-login`);

      const response = await fetch(`${API_BASE_URL}/auth/last-login`, {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
      });

      console.log('📥 Last login response status:', response.status);

      if (!response.ok) {
        if (response.status === 404) {
          console.warn('⚠️ Last login endpoint not available');
          return null;
        }
        const error = await response.json();
        throw new Error(error.detail || error.message || 'Không thể lấy thông tin đăng nhập lần cuối');
      }

      const result = await response.json();
      console.log('✅ Last login info fetched:', result);

      return result as LastLogin;
    } catch (error) {
      console.error('❌ Get last login error:', error);
      return null;
    }
  }
}

File: Downloads/backend/src/services/auth_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException, status
from datetime import datetime, timedelta
from uuid import uuid4
import logging

from src.models.user import User, UserProfile, ActivityLog
from src.schemas.user import UserRegister, UserLogin, Token, UserResponse
from src.utils.security import (
    get_password_hash,
    verify_password,
    create_access_token
)
from src.utils.timezone import get_vietnam_now
from src.config.settings import get_settings

settings = get_settings()

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.INFO)


class AuthService:
    def __init__(self, db: Session):
        self.db = db

    def register_user(self, user_data: UserRegister, ip_address: str | None = None, user_agent: str | None = None) -> Token:
        """
        Đăng ký người dùng mới
        - Email được normalize thành lowercase
        - Username lưu nguyên vẹn nhưng kiểm tra case-insensitive
        """
        logger.info(f"=== REGISTER REQUEST RECEIVED ===")
        logger.info(f"Email: {user_data.email}, Username: {user_data.username}")
        logger.info(f"IP Address: {ip_address}, User Agent: {user_agent}")

        # Normalize email thành lowercase
        email_normalized = user_data.email.lower().strip()
        username_input = user_data.username.strip()

        # Kiểm tra email đã tồn tại (case-insensitive)
        existing_email = self.db.query(User).filter(
            User.email == email_normalized
        ).first()
        if existing_email:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Email đã được đăng ký"
            )

        # Kiểm tra username đã tồn tại (case-insensitive)
        existing_username = self.db.query(User).filter(
            User.username.ilike(username_input)
        ).first()
        if existing_username:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Username đã tồn tại"
            )

        # Kiểm tra password match
        if user_data.password != user_data.confirm_password:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Mật khẩu không khớp"
            )

        # Validate password strength
        if not self._validate_password_strength(user_data.password):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Mật khẩu phải có ít nhất 8 ký tự, bao gồm chữ hoa, chữ thường, số và ký tự đặc biệt"
            )

        try:
            # Tạo user mới
            now = get_vietnam_now()
            new_user = User(
                id=uuid4(),
                email=email_normalized,  # Lưu email normalized
                username=username_input,  # Lưu username nguyên vẹn
                password_hash=get_password_hash(user_data.password),
                role='user',
                status='active',
                email_verified=False,
                created_at=now,
                last_login=now
            )

            self.db.add(new_user)
            self.db.flush()
            logger.info(f"✅ User created: {new_user.id}")

            # Tạo user profile
            new_profile = UserProfile(
                id=uuid4(),
                user_id=new_user.id,
                display_name=username_input,
                language='vi',
                default_currency='USDT',
                notify_email=True,
                notify_push=True
            )

            self.db.add(new_profile)
            logger.info(f"✅ Profile created: {new_profile.id}")

            # Ghi activity log
            activity = ActivityLog(
                id=uuid4(),
                user_id=new_user.id,
                action='register',
                details=f"Đăng ký tài khoản mới. Email: {email_normalized}, Username: {username_input}",
                created_at=now
            )
            self.db.add(activity)

            self.db.commit()
            logger.info(f"✅ Activity logged")

            # Tạo token
            access_token = create_access_token(
                data={
                    "sub": new_user.email,
                    "user_id": str(new_user.id)
                }
            )

            logger.info(f"✅ Token created, registration successful")

            return Token(
                access_token=access_token,
                token_type="bearer",
                expires_in=settings.access_token_expire_minutes * 60,
                user=UserResponse.model_validate({
                    "id": str(new_user.id),
                    "email": str(new_user.email),
                    "username": str(new_user.username),
                    "role": str(new_user.role),
                    "status": str(new_user.status),
                    "email_verified": bool(new_user.email_verified),
                    "created_at": new_user.created_at,
                    "last_login": new_user.last_login
                })
            )
        except Exception as e:
            self.db.rollback()
            logger.error(f"❌ Registration error: {str(e)}")
            raise

    def login_user(self, user_data: UserLogin, ip_address: str | None = None, user_agent: str | None = None) -> Token:
        """
        Đăng nhập người dùng
        - Email được normalize thành lowercase để kiểm tra
        """
        logger.info(f"=== LOGIN REQUEST RECEIVED ===")
        logger.info(f"Email: {user_data.email}")
        logger.info(f"IP Address: {ip_address}, User Agent: {user_agent}")

        # Normalize email thành lowercase
        email_normalized = user_data.email.lower().strip()

        try:
            # Tìm user bằng email normalized
            user = self.db.query(User).filter(User.email == email_normalized).first()

            if not user:
                logger.warning(f"❌ User not found: {email_normalized}")
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Email hoặc mật khẩu không chính xác"
                )

            # Verify password
            if not verify_password(user_data.password, str(user.password_hash)):
                logger.warning(f"❌ Invalid password for: {email_normalized}")
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Email hoặc mật khẩu không chính xác"
                )

            if str(user.status) != 'active':
                logger.warning(f"❌ Account locked: {email_normalized}")
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="Tài khoản của bạn đã bị khóa"
                )

            # Cập nhật last_login
            new_last_login = get_vietnam_now()
            self.db.query(User).filter(User.id == user.id).update(
                {User.last_login: new_last_login}
            )

            # Ghi activity log
            activity = ActivityLog(
                id=uuid4(),
                user_id=user.id,
                action='login',
                details=f"Đăng nhập thành công. Email: {user.email}",
                created_at=get_vietnam_now()
            )
            self.db.add(activity)
            self.db.commit()

            logger.info(f"✅ Login successful: {email_normalized}")

            # Tạo token
            access_token = create_access_token(
                data={
                    "sub": user.email,
                    "user_id": str(user.id)
                }
            )

            return Token(
                access_token=access_token,
                token_type="bearer",
                expires_in=settings.access_token_expire_minutes * 60,
                user=UserResponse.model_validate({
                    "id": str(user.id),
                    "email": str(user.email),
                    "username": str(user.username),
                    "role": str(user.role),
                    "status": str(user.status),
                    "email_verified": bool(user.email_verified),
                    "created_at": user.created_at,
                    "last_login": new_last_login
                })
            )
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"❌ Login error: {str(e)}")
            self.db.rollback()
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Lỗi server khi đăng nhập"
            )

    def logout_user(self, user_id: str) -> dict:
        """
        Đăng xuất người dùng (chỉ ghi log, JWT không có blacklist)
        """
        user = self.db.query(User).filter(User.id == user_id).first()

        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Người dùng không tồn tại"
            )

        # Ghi activity log
        activity = ActivityLog(
            id=uuid4(),
            user_id=user.id,
            action='logout',
            details=f"Đăng xuất. Email: {user.email}",
            created_at=get_vietnam_now()
        )
        self.db.add(activity)
        self.db.commit()

        return {"message": "Đăng xuất thành công"}

    def change_password(self, current_user: User, password_data) -> dict:
        """
        Thay đổi mật khẩu
        """
        logger.info(f"=== CHANGE PASSWORD REQUEST ===")
        logger.info(f"User: {current_user.email}")

        try:
            # Verify current password
            if not verify_password(password_data.current_password, str(current_user.password_hash)):
                logger.warning(f"❌ Invalid current password: {current_user.email}")
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Mật khẩu hiện tại không chính xác"
                )
            
            # Check new password match
            if password_data.new_password != password_data.confirm_new_password:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Mật khẩu mới không khớp"
                )
            
            # Validate password strength
            if not self._validate_password_strength(password_data.new_password):
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Mật khẩu phải có ít nhất 8 ký tự, bao gồm chữ hoa, chữ thường, số và ký tự đặc biệt"
                )
            
            # Update password
            new_hash = get_password_hash(password_data.new_password)
            self.db.query(User).filter(User.id == current_user.id).update(
                {User.password_hash: new_hash}
            )
            
            # Log activity
            activity = ActivityLog(
                id=uuid4(),
                user_id=current_user.id,
                action='password_change',
                details="Thay đổi mật khẩu thành công",
                created_at=get_vietnam_now()
            )
            self.db.add(activity)
            self.db.commit()
            
            logger.info(f"✅ Password changed successfully: {current_user.email}")
            
            return {"message": "Thay đổi mật khẩu thành công"}
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"❌ Change password error: {str(e)}")
            self.db.rollback()
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Lỗi server khi thay đổi mật khẩu"
            )

    @staticmethod
    def _validate_password_strength(password: str) -> bool:
        """
        Validate password strength
        - Tối thiểu 8 ký tự
        - Có ít nhất 1 chữ hoa
        - Có ít nhất 1 chữ thường
        - Có ít nhất 1 số
        - Có ít nhất 1 ký tự đặc biệt
        """
        if len(password) < 8:
            return False

        has_upper = any(c.isupper() for c in password)
        has_lower = any(c.islower() for c in password)
        has_digit = any(c.isdigit() for c in password)
        has_special = any(not c.isalnum() for c in password)

        return has_upper and has_lower and has_digit and has_special

